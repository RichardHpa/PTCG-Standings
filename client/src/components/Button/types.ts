import type { ReactNode, ComponentPropsWithoutRef } from 'react';
import type { Colors } from 'types/colors';

export interface ButtonProps extends ComponentPropsWithoutRef<'button'> {
  children: ReactNode;
  color?: Colors;
  variant?: 'contained' | 'outlined' | 'ghost';
  size?: 'xs' | 'sm' | 'base' | 'lg' | 'xl';
}
