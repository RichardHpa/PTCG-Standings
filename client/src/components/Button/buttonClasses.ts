import { tw } from 'utils/tailwindClassName';

export const buttonClasses = {
  base: tw`inline-flex rounded-lg text-center font-medium`,
  variant: {
    contained: {
      base: tw``,
      primary: {
        light: tw`bg-primary text-white hover:bg-primary-hover`,
        dark: tw`text-white dark:bg-primary-dark dark:hover:bg-primary-hoverDark`,
      },
      secondary: {
        light: tw`bg-secondary text-white hover:bg-secondary-hover`,
        dark: tw`text-white dark:bg-secondary-dark dark:hover:bg-secondary-hoverDark`,
      },
      alternative: {
        light: tw`bg-white text-gray-900 hover:bg-gray-100`,
        dark: tw`dark:bg-gray-800 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white`,
      },
    },
    outlined: {
      base: tw`outline outline-1 hover:text-white`,
      primary: {
        light: tw`text-primary outline-primary hover:bg-primary hover:outline-primary`,
        dark: tw`dark:text-primary-dark dark:outline-primary-dark hover:dark:bg-primary-dark`,
      },
      secondary: {
        light: tw`text-secondary outline-secondary hover:bg-secondary hover:outline-secondary`,
        dark: tw`dark:text-secondary-dark dark:outline-secondary-dark hover:dark:bg-secondary-dark`,
      },
      alternative: {
        light: tw`text-gray-900 outline-gray-900 hover:bg-gray-900 hover:text-white`,
        dark: tw`dark:text-gray-400 dark:outline-gray-700 dark:hover:bg-gray-700 dark:hover:text-white`,
      },
    },
    ghost: {
      base: tw``,
      primary: {
        light: tw`text-primary hover:bg-primary-hoverLight hover:text-primary`,
        dark: tw`dark:text-primary-dark dark:hover:bg-primary-hover dark:hover:bg-primary-hoverDark dark:hover:text-white`,
      },
      secondary: {
        light: tw`text-secondary hover:bg-secondary-hoverLight hover:text-secondary`,
        dark: tw`dark:text-secondary-dark dark:hover:bg-secondary-hover dark:hover:bg-secondary-hoverDark dark:hover:text-white`,
      },
      alternative: {
        light: tw`text-gray-900 hover:bg-gray-100`,
        dark: tw`dark:text-gray-400 dark:hover:bg-gray-700`,
      },
    },
  },
  size: {
    xs: tw`px-3 py-2 text-xs`,
    sm: tw`px-3 py-2 text-sm`,
    base: tw`px-5 py-2.5 text-sm`,
    lg: tw`px-5 py-3 text-base`,
    xl: tw`px-6 py-3.5 text-base`,
  },
};
