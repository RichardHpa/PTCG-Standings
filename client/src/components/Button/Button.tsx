import clsx from 'clsx';

import { buttonClasses } from './buttonClasses';

import type { FC } from 'react';
import type { ButtonProps } from './types';

export const Button: FC<ButtonProps> = ({
  children,
  color = 'primary',
  variant = 'contained',
  size = 'base',
  ...props
}) => {
  return (
    <button
      type="button"
      {...props}
      className={clsx(
        buttonClasses.base,
        buttonClasses.size[size],
        buttonClasses.variant[variant]['base'],
        buttonClasses.variant[variant][color].light,
        buttonClasses.variant[variant][color].dark,
      )}
    >
      {children}
    </button>
  );
};
