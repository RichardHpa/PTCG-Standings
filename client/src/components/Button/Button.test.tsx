import { describe, expect, test, vi } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Button } from './Button';

describe('Button', () => {
  test('renders default button', () => {
    render(<Button>Default</Button>);
    const button = screen.getByRole('button', { name: /Default/i });
    expect(button).toBeInTheDocument();
  });

  test('renders Button with onClick', async () => {
    const onClick = vi.fn();
    render(<Button onClick={onClick}>Test</Button>);
    userEvent.click(screen.getByRole('button'));
    await waitFor(() => expect(onClick).toHaveBeenCalled());
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  test('renders Button with type', () => {
    render(<Button type="submit">Test</Button>);
    expect(screen.getByRole('button')).toHaveAttribute('type', 'submit');
  });

  test.each([['contained'], ['outlined'], ['ghost']] as const)(
    'renders Button with variant %s',
    variant => {
      const { container } = render(
        <Button variant={variant}>{variant}</Button>,
      );
      expect(container).toMatchSnapshot();
    },
  );
});
