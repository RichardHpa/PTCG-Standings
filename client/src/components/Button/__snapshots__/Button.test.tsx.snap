// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders Button with variant contained 1`] = `
<div>
  <button
    class="border-1 inline-flex border text-center font-medium rounded-md px-4 py-2 text-sm border-transparent bg-primary text-white hover:bg-primary-hover text-white dark:bg-primary-dark dark:hover:bg-primary-hoverDark"
    type="button"
  >
    contained
  </button>
</div>
`;

exports[`Button > renders Button with variant ghost 1`] = `
<div>
  <button
    class="border-1 inline-flex border text-center font-medium rounded-md px-4 py-2 text-sm border-transparent text-primary hover:bg-primary-hoverLight hover:text-primary dark:text-primary-dark dark:hover:bg-primary-hover dark:hover:bg-primary-hoverDark dark:hover:text-white"
    type="button"
  >
    ghost
  </button>
</div>
`;

exports[`Button > renders Button with variant outlined 1`] = `
<div>
  <button
    class="border-1 inline-flex border text-center font-medium rounded-md px-4 py-2 text-sm hover:text-white border-primary text-primary hover:border-primary hover:bg-primary dark:border-primary-dark dark:text-primary-dark hover:dark:bg-primary-dark"
    type="button"
  >
    outlined
  </button>
</div>
`;
